src/lib.rs

```diff
@@ -213,108 +206,251 @@ impl Build {
+                "ws_engine",
+                "ws_listener",
+                // "wss_address", // requires gnutls
+                // "wss_engine", // requires gnutls
+                "xpub",
+                "xsub",
+                "zap_client",
+                "zmq_utils",
+                "zmq",
+                "zmtp_engine",
+            ],
+        );
+
+        add_c_sources(
+            &mut build,
+            path.join("external/sha1"),
+            &[
+                "sha1.c"
+            ],
+        );
+
+        //build.define("ZMQ_WIN32_WINNT_DEFAULT", "1");
+        build.define("ZMQ_USE_CV_IMPL_STL11", "1");
+        build.define("ZMQ_BUILD_DRAFT_API", "1");
+        build.define("ZMQ_STATIC", "1");
+        build.define("ZMQ_USE_BUILTIN_SHA1", "1");
+
+        build.define("ZMQ_HAVE_WS", "1");
+
+        //build.define("ZMQ_HAVE_IPC", "1");  // IPC doesn't work on windows with `select`
+
+        //build.define("ZMQ_USE_GNUTLS", "1");
+
+        // if self.enable_draft {
+        //     build.define("ENABLE_DRAFTS", "ON");
+        // } else {
+        //     build.define("ENABLE_DRAFTS", "OFF");
+        // }
+
+        // if self.enable_curve {
+        //     build.define("ENABLE_CURVE", "ON");
+        // } else {
+        //     build.define("ENABLE_CURVE", "OFF");
+        // }
+
+        // if self.build_debug {
+        //     build.define("CMAKE_BUILD_TYPE", "Debug");
+        // } else {
+        //     build.define("CMAKE_BUILD_TYPE", "Release");
+        // }
+
+        // if self.perf_tool {
+        //     build.define("WITH_PERF_TOOL", "ON");
+        // } else {
+        //     build.define("WITH_PERF_TOOL", "OFF");
+        // }
+
+        if let Some(_) = &self.libsodium {
+            build.define("ZMQ_USE_LIBSODIUM", "1");
         }
 
-        if self.enable_curve {
-            config.define("ENABLE_CURVE", "ON");
-        } else {
-            config.define("ENABLE_CURVE", "OFF");
-        }
-
-        if self.build_debug {
-            config.define("CMAKE_BUILD_TYPE", "Debug");
-        } else {
-            config.define("CMAKE_BUILD_TYPE", "Release");
-        }
-
-        if self.perf_tool {
-            config.define("WITH_PERF_TOOL", "ON");
-        } else {
-            config.define("WITH_PERF_TOOL", "OFF");
-        }
+        let mut create_platform_hpp_shim = || {
+            // https://cmake.org/cmake/help/latest/command/configure_file.html
+            // TODO: Replace `#cmakedefine` with the appropriate `#define`
+            // let _platform_file =
+            //     std::fs::read_to_string(path.join("builds/cmake/platform.hpp.in"))
+            //         .unwrap();
+
+            // TODO: Write these to a neat directory inside `out`
+            let out_includes = PathBuf::from(std::env::var("OUT_DIR").unwrap());
+            // Write out an empty platform file: defines will be set through cc directly
+            std::fs::write(out_includes.join("platform.hpp"), "").unwrap();
```

Comment:
Don't forget to flush after writing, otherwise this can create a race on some systems.

Reply:
Done.